src/clojure/nrepl/middleware/interruptible_eval.clj

```diff
@@ -47,88 +47,65 @@
-           :need-prompt (constantly false)
-           ;; TODO: pretty-print?
-           :print (fn [v]
-                    (reset! bindings (assoc (capture-thread-bindings)
-                                            #'*3 *2
-                                            #'*2 *1
-                                            #'*1 v))
-                    (.flush ^Writer err)
-                    (.flush ^Writer out)
-                    (reset! session (maybe-restore-original-ns @bindings))
-                    (t/send transport (response-for msg
-                                                    {:value v
-                                                     :ns (-> *ns* ns-name str)})))
-           ;; TODO: customizable exception prints
-           :caught (fn [e]
-                     (let [root-ex (#'clojure.main/root-cause e)
-                           previous-cause (.getCause e)]
-                       ;; Check if the root cause or previous cause of the exception
-                       ;; is a ThreadDeath exception. In case the exception is a
-                       ;; CompilerException, the root cause is not returned by
-                       ;; the root-cause function, so we check the previous cause
-                       ;; instead.
-                       (when-not (or (instance? ThreadDeath root-ex)
-                                     (instance? ThreadDeath previous-cause))
-                         (reset! bindings (assoc (capture-thread-bindings) #'*e e))
-                         (reset! session (maybe-restore-original-ns @bindings))
-                         (t/send transport (response-for msg {:status :eval-error
-                                                              :ex (-> e class str)
-                                                              :root-ex (-> root-ex class str)}))
-                         (clojure.main/repl-caught e)))))
-          (finally
-            (.flush ^Writer out)
-            (.flush ^Writer err)))))
-    (maybe-restore-original-ns @bindings)))
+      (try
+        (clojure.main/repl
+         :eval (if eval (find-var (symbol eval)) clojure.core/eval)
+         ;; clojure.main/repl paves over certain vars even if they're already thread-bound
+         :init #(let [bindings
+                      (-> (get-thread-bindings)
+                          (into @session)
+                          (cond-> explicit-ns (assoc #'*ns* explicit-ns)
+                                  file (assoc #'*file* file)))]
+                  (pop-thread-bindings)
+                  (push-thread-bindings bindings))
+         :read (if (string? code)
+                 (let [reader (source-logging-pushback-reader code line column)]
+                   #(read {:read-cond :allow :eof %2} reader))
+                 (let [code (.iterator ^Iterable code)]
+                   #(or (and (.hasNext code) (.next code)) %2)))
+         :prompt #(reset! session (maybe-restore-original-ns (capture-thread-bindings)))
```

Comment:
That's called after init, right?

Reply:
it's called after init and before read, it means it's called before EOF too.
As a bonus we can leave the `*N` vars management to clojure.main/repl!